File: solutions/go/luhn/1/luhn.go
package luhn
import ( 
    //"fmt"
    "unicode"
    )
func Valid(id string) bool {
	var double bool = false    
    var digits []rune = []rune(id)
    var length int = len(digits)
    var sum int = 0
    if length == 1 && digits[0] == '0' {
        return false
    }
    for i := length-1; i >= 0; i-- {
        var v rune = digits[i]
        if unicode.IsDigit(v) {
            var digit int = int(v - '0')
            if double {
                digit *= 2
                if digit > 9 {
                    digit -= 9
                }
                double = false
            } else {
            	double = true
            }        
            sum += digit
        } else if v == ' ' && length == 2 {
        	return false
        } else if v == ' ' {
			continue
		} else {
        	return false
        }
    }
    return sum % 10 == 0
}

File: solutions/ruby/luhn/1/luhn.rb
=begin
Write your code for the 'Luhn' exercise in this file. Make the tests in
`luhn_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/luhn` directory.
=end

class Luhn
  def self.valid?(number)
    return false if number.length < 2 || number.scan(/[^0-9 ]+/).length > 0 || number.scan(/[0-9]/).length < 2
    number.scan(/[0-9]/).reverse.map.each_with_index {|val, index| index.odd? && val.to_i < 9 ? self.multiply(val.to_i) : val.to_i }.sum % 10 == 0
  end
  
  def self.multiply(num)
      v = num * 2
      v > 9 ? v - 9 : v
  end
end